File: frontend/lib/model/card.abstract.dart
import 'package:frontend/model/subcard.abstract.dart';
import 'package:frontend/model/utils.dart';
import 'package:json_annotation/json_annotation.dart';

abstract class CardAbstract extends EditableItemAbstract {
  bool _isCompleted = false;

  bool get isCompleted => _isCompleted;

  set isCompleted(bool value) {
    _isCompleted = value;
    if (value) {
      isExpanded = false;
    }
  }

  @JsonKey(includeFromJson: false, includeToJson: false)
  bool isExpanded = true;

  @JsonKey(includeFromJson: false, includeToJson: false)
  bool canAddItems = true;

  CardAbstract();

  @override
  List<FieldDescriptor> getEditableFields();
}

File: react-native-expo/models/abstract/CardAbstract.ts
import { EditableItemAbstract } from './EditableItemAbstract';
import { FieldDescriptor } from '../utils';

/**
 * Base abstract class for all card items
 * Equivalent to Dart's CardAbstract
 */
export abstract class CardAbstract extends EditableItemAbstract {
  private _isCompleted: boolean = false;
  isExpanded: boolean = true;
  canAddItems: boolean = true;

  constructor() {
    super();
  }

  /**
   * Completion status with automatic expansion control
   */
  get isCompleted(): boolean {
    return this._isCompleted;
  }

  set isCompleted(value: boolean) {
    this._isCompleted = value;
    if (value) {
      this.isExpanded = false;
    }
  }

  /**
   * Abstract method that must be implemented by concrete classes
   */
  abstract getEditableFields(): FieldDescriptor[];
}
